src/components/TodoItem.tsx

```diff
@@ -0,0 +1,169 @@
+          const newTodos = [...currentTodos];
+          const index = newTodos.findIndex(td => td.id === updatedTodo.id);
+
+          newTodos.splice(index, 1, newTodo);
+
+          return newTodos;
+        });
+      })
+      .catch(() => setErrorMessage('Unable to update a todo'))
+      .finally(() => setToggled(false));
+  };
+
+  const handleDoubleClick = () => {
+    setIsEditing(true);
+  };
+
+  const handleSubmit = (event: React.FormEvent) => {
+    event.preventDefault();
+    setIsEditing(false);
+
+    if (newTitle.trim().length === 0) {
+      handleDeleteTodo(todo.id);
+    } else if (newTitle === todo.title) {
+      setIsEditing(false);
+    } else {
+      setSelectedTodo(todo);
+
+      updateTodo({ ...todo, title: newTitle })
+        .then(newTodo => {
+          setTodos(currentTodos => {
+            const newTodos = [...currentTodos];
+            const index = newTodos.findIndex(td => td.id === newTodo.id);
+
+            newTodos.splice(index, 1, newTodo);
+
+            return newTodos;
+          });
+        })
+        .catch((error) => {
+          setErrorMessage('Unable to update a todo');
+          setNewTitle(todo.title);
+          throw error;
+        })
+        .finally(() => {
+          setIsEditing(false);
+          setSelectedTodo(null);
+        });
+    }
+  };
+
+  const handleKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+    if (event.key === 'Escape') {
+      setIsEditing(false);
+      setNewTitle(todo.title);
+    }
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={classNames('todo', {
+        completed: todo.completed,
+      })}
+      onDoubleClick={handleDoubleClick}
+      key={todo.id}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={todo.completed}
+          onChange={() => toggleCompletedTodo(todo)}
+        />
+      </label>
+
+      {isEditing
+        ? (
+          <form onSubmit={handleSubmit}>
+            <input
+              data-cy="TodoTitleField"
+              type="text"
+              className="todo__title-field"
+              placeholder="Empty todo will be deleted"
+              ref={editTitleFieldRef}
+              value={newTitle}
+              onChange={editTitle}
+              onBlur={handleSubmit}
+              onKeyUp={handleKeyUp}
+            />
+          </form>
+        )
+        : (
+          <>
+            <span data-cy="TodoTitle" className="todo__title">
+              {selectedTodo ? newTitle : todo.title}
+            </span>
+
+            <button
+              type="button"
+              className="todo__remove"
+              data-cy="TodoDelete"
+              onClick={() => handleDeleteTodo(todo.id)}
+            >
+              ×
+            </button>
+          </>
+        )}
+
+      {(todo.id === 0 || isDeliting || toggled || selectedTodo) && (
```

Comment:
currently, you don't show the loader for batch actions (toggle all, clear completed), and use different variables to show the loader. You may create a 1 separate on a higher level with `loadingTodoIds` array (empty by default). Once you do any action for todo add its id to the state mentioned above. When the action is done - remove the id. Inside this component check if in those array exists id of current todo - if yes - show loader (and gray background)